### ISSUE TITLE
Running `npm prune` after `npm install` causes `RangeError` and npm to crash

### ISSUE THREAD

**Issue by ntwb** [npm/npm#17419]
#### I'm opening this issue because:

  - [x] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

After running `npm install` and then running `npm prune` npm crashes

#### How can the CLI team reproduce the problem?

1. Ensure you are using Node.js `8.1.2` and npm `5.0.3`
2. Run `npm install`
3. Run `npm prune`

### supporting information:

 - `npm -v` prints: `5.0.3`
 - `node -v` prints: `8.1.2`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: macOS
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

Travis CI job showing the issue: https://travis-ci.org/WordPress-Coding-Standards/eslint-plugin-wordpress/jobs/246573344#L177 :
```shell
26.05s$ npm install
> fsevents@1.1.2 install /home/travis/build/WordPress-Coding-Standards/eslint-plugin-wordpress/node_modules/fsevents
> node install
npm notice created a lockfile as package-lock.json. You should commit this file.
added 817 packages in 25.569s
8.52s$ npm prune
Unhandled rejection RangeError: Maximum call stack size exceeded
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install/node.js:33:25)
    at /home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install/node.js:36:14
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install/node.js:33:25)
    at /home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install/node.js:36:14
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install/node.js:33:25)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install.js:377:5)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install.js:379:19)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install.js:379:19)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.2/lib/node_modules/npm/lib/install.js:379:19)
    at Array.forEach (native)
npm ERR! cb() never called!
npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2017-06-24T15_45_53_557Z-debug.log
The command "npm prune" failed and exited with 1 during .
Your build has been stopped.
```

The resulting `debug.log`: 
```text
0 info it worked if it ends with ok
1 verbose cli [ '/Users/netweb/.nvm/versions/node/v8.1.2/bin/node',
1 verbose cli   '/Users/netweb/.nvm/versions/node/v8.1.2/bin/npm',
1 verbose cli   'prune' ]
2 info using npm@5.0.3
3 info using node@v8.1.2
4 verbose npm-session aafac1926fe00481
5 silly install loadCurrentTree
6 silly install readLocalPackageData
7 error cb() never called!
8 error This is an error with npm itself. Please report this error at:
9 error <https://github.com/npm/npm/issues>
```

Possibly related to or duplicate of #16853

**Comment by ntwb** [comment]
I'd hoped this was fixed in npm 5.1.0, it appears not:

> https://travis-ci.org/WordPress-Coding-Standards/eslint-plugin-wordpress/jobs/251079507
```shell
$ npm install -g npm@latest
/home/travis/.nvm/versions/node/v8.1.3/bin/npm -> /home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/bin/npm-cli.js
+ npm@5.1.0
added 27 packages, removed 16 packages and updated 47 packages in 9.795s
28.48s$ npm install 
Unhandled rejection RangeError: Maximum call stack size exceeded
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install/node.js:36:25)
    at /home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install/node.js:39:14
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install/node.js:36:25)
    at /home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install/node.js:39:14
    at Array.forEach (native)
    at exports.create (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install/node.js:36:25)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install.js:376:5)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install.js:378:19)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install.js:378:19)
    at Array.forEach (native)
    at normalizeTree (/home/travis/.nvm/versions/node/v8.1.3/lib/node_modules/npm/lib/install.js:378:19)
    at Array.forEach (native)
npm ERR! cb() never called!
npm ERR! This is an error with npm itself. Please report this error at:
npm ERR!     <https://github.com/npm/npm/issues>
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/travis/.npm/_logs/2017-07-07T09_05_34_797Z-debug.log
```

**Comment by ntwb** [comment]
Turns out this is due to _[local link sources](https://docs.npmjs.com/all#version-1)_

i.e I had the following dev dependency: `"eslint-plugin-wordpress": "file:.",`

Running `npm install` when using `file:.` with npm 5.0/5.1 results in npm crashing per docs above
